### ISSUE TITLE
Instances for Enter are secretly overlapping

### ISSUE THREAD

**Issue by fizruk** [haskell-servant/servant#734]
Turns out these two instances secretly overlap:

```haskell
instance (Enter b arg ret) => Enter (a -> b) arg (a -> ret)
instance Enter (m a) (m :~> n) (n a)
```

It happens only when `m ~ n` and I suspect that's why no one noticed that so far.

In the "real world" this results in non-descriptive type errors like this:

```
    • Couldn't match type ‘App’ with ‘(->) (Maybe Param)’
        arising from a use of ‘enter’
    • In the expression: enter (Nat id :: App :~> App) server
      In an equation for ...
```

This prevents me from applying `App :~> App` transformation to handlers directly.
Currently I work around this by using a pair of transformation:

```haskell
Nat IdentityT :: App :~> IdentityT App

Nat (f . runIdentityT) :: IdentityT App :~> App
```

instead of just

```haskell
Nat f :: App :~> App
```

Unfortunately I don't have a minimal failing test case at this moment. For instance, this works just fine:

```
>>> :set -XDataKinds
>>> :set -XTypeOperators
>>> import Servant
>>> type Endpoint1 = QueryParam "param" Int :> Get '[JSON] NoContent
>>> type Endpoint2 = QueryParam "param" String :> Get '[JSON] NoContent
>>> type API = Endpoint1 :<|> Endpoint2
>>> data App a
>>> type S layout = ServerT layout App
>>> server :: S API; server = undefined
>>> :t enter (Nat id :: App :~> App) server
enter (Nat id :: App :~> App) server
  :: (Maybe Int -> App NoContent)
     :<|> (Maybe [Char] -> App NoContent)
```

Perhaps there's something extra in my project's API that prevents GHC to infer types correctly.

**Comment by phadej** [comment]
Hmm the fact that pair of transformations works kind of rules out that you have non `App` endpoint impl. Weird.

**Comment by fizruk** [comment]
Apparently it works differently in REPL.
Here's a minimal example:

```haskell
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeOperators #-}

import Servant

data App a

f :: App :~> App
f = Nat id

server :: App Int :<|> (String -> App Bool)
server = undefined

server' :: App Int :<|> (String -> App Bool)
server' = enter f server
```

Here's what happens if you try load it in GHCi:

```
>>> :l ~/temp/servant-enter.hs
[1 of 1] Compiling Main             ( /Users/nickolaykudasov/temp/servant-enter.hs, interpreted )

/Users/nickolaykudasov/temp/servant-enter.hs:15:11: error:
    • Couldn't match type ‘App’ with ‘(->) String’
        arising from a use of ‘enter’
    • In the expression: enter f server
      In an equation for ‘server'’: server' = enter f server
Failed, modules loaded: none.
```

**Comment by fizruk** [comment]
Note that using alternative (`:<|>`) is critical for the example to fail:

```haskell
-- | Compilation fails for this server type.
server :: App Int :<|> (String -> App Bool)

-- | Compilation works for this one though.
server :: String -> App Bool
```

**Comment by alpmestan** [comment]
That's not really surprising. `Enter` sucks, if anyone wants to research a better way to accomplish what it does, I'm definitely up for sponsoring the beers required to go through the process of replacing it =)

**Comment by phadej** [comment]
```haskell
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeOperators #-}

import Servant

data App a

f :: App :~> App
f = Nat id

server :: App Int :<|> (String -> App Bool)
server = undefined

server' :: App Int :<|> (String -> App Bool)
server' = enter f server
```

Dunno if it really works, but @fizruk's example compiles

**Comment by phadej** [comment]
```haskell

type family Entered m n api where
    Entered m n (m a)            = n a
    Entered m n (a -> api)       = a -> Entered m n api
    Entered m n (api1 :<|> api2) = Entered m n api1 :<|> Entered m n api2

class ( Entered m n typ ~ ret
      , Entered n m ret ~ typ
      )
    => Enter typ m n ret
  where
    enter :: m :~> n -> typ -> ret

-- **  Servant combinators

instance
    ( Entered m1 n1 (typ1 :<|> typ2) ~ (ret1 :<|> ret2)
    , Entered n1 m1 (ret1 :<|> ret2) ~ (typ1 :<|> typ2)
         , Enter typ1 m1 n1 ret1, Enter typ2 m2 n2 ret2
         , m1 ~ m2, n1 ~ n2
         ) => Enter (typ1 :<|> typ2) m1 n1 (ret1 :<|> ret2) where
    enter e (a :<|> b) = enter e a :<|> enter e b

instance
    ( Entered m n (a -> typ) ~ (a -> ret)
    , Entered n m (a -> ret) ~ (a -> typ)
    , Enter typ m n ret
    )
    => Enter (a -> typ) m n (a -> ret)
  where
    enter arg f a = enter arg (f a)

-- ** Useful instances

instance Enter (m a) m n (n a) where
    enter (NT f) = f
```

I'm not sure this really works, but @fizruk 's example type-checks

**Comment by fizruk** [comment]
Honestly, I start to not like the way we combine our API with `:>` and `:<|>`. This simple at first sight approach leads to multiple design problems the biggest of which are probably `Enter` and `ServerT`. In short, handling `:>` and `:<|>` looks to me like handling `GHC.Generics` and I think there is a nicer solution (something more like `generics-sop`) for specifying type-level APIs. However, as much as I would love to investigate alternative API representations, I simply don't have time right now.

**Comment by phadej** [comment]
@fizruk have you seen #713 and http://oleg.fi/gists/posts/2017-03-15-solution-to-servant-type-indentation-problem.html ?

**Comment by arianvp** [comment]
I would prefer to get rid of the arbitrary nesting with ` :<|>`  it is ugly.  I'd prefer a list of API endpoints instead. Seems similar to what @phadej is coming up with.  I just that too for `https://github.com/arianvp/servis`, my idris port.

First of all, it will make things like enter a lot more simple, as it would simply be a `map` over the list.

Second of all,  we can do cool stuff like  define handlers out of order  and use some kind of `ElemInside` `Constraint` in the `HasServer` instances. This would mean your server implementation does not have to exactly match the nesting and order of how your API is defined. 

Furthermore, we can easily implement some helpful dianostics.  Like, you could add a type level proof that no routes overlap. (Again, see Servis).


So yeh, i'm totally for moving to a more some-of-products like structure

**Comment by arianvp** [comment]
OTOH, this would perhaps cause some problems for modularity.

**Comment by kosmikus** [comment]
Yes, my concern is modularity. I'm not opposed to using lists in some way, and they certainly look somewhat nicer for APIs that are specified completely in one place. But I think we still need an operator like `:<|>` that just combines two API fragments.

However, in principle, we can compute with type-level APIs, and the way we write them does not have to correspond to the way we represent them internally when we make computations. Computing normal forms would be easier though if we could make the type-level grammars more precise, but that still seems to be blocked by the absence of user-defined open kinds.

**Comment by fizruk** [comment]
@phadej yes, that's a step in the right direction and that approach might probably solve `Enter`. However we also need to solve the situation with `:>`. IMO, the worst thing about `:>` is that it mixes methods and parameters, resulting in an ugly `ServerT` we have today.

@arianvp 

> This would mean your server implementation does not have to exactly match the nesting and order of how your API is defined.

This one sounds like ambiguous types pain :)

> you could add a type level proof that no routes overlap

It is possible for Servant, and it's been discussed before (probably on IRC). Just no one has taken this task upon himself yet, I guess.

> some-of-products like structure

I suggest there's more to API than sum-of-products. I think API is more like sum of endpoints. Currently we mix all endpoint's properties (path segments, parameters, methods) together, but it does not seem to work out nice. I think we have to make endpoints more structured.

P.S. We probably should move discussion elsewhere.

**Comment by alpmestan** [comment]
> Honestly, I start to not like the way we combine our APIs with :> and :<|>. This simple at first sight approach leads to multiple design problems the biggest of which are probably Enter and ServerT. In short, handling :> and :<|> looks to me like handling GHC.Generics and I think there is a nicer solution (something more like generics-sop) for specifying type-level APIs. However, as much as I would love to investigate alternative API representations, I simply don't have time right now.

@fizruk It feels to me this is simply because we have never really figured out what kind of structure our API types give us. There is a sum-of-products-y feel to it, but I don't think it's a match made in heaven either.

> OTOH, this would perhaps cause some problems for modularity.

@arianvp I already thought about this, back in the beginning, and it does break modularity, unless you combine-and-flatten at the same time in some way, but I don't like that idea one bit.

> Computing normal forms would be easier though if we could make the type-level grammars more precise, but that still seems to be blocked by the absence of user-defined open kinds.

@kosmikus That's exactly my feeling. We've been fighting the lack of user-defined open kinds since the beginning. I realized that only when we wrote the paper, during our discussions. Looks like we're stuck with that fight for the foreseeable future... I do however think that normal forms for API types have a lot to bring to the table, my hacky `Canonicalize` was an experiment in that direction and I've always felt there was more to that.

> I suggest there's more to API than sum-of-products. I think API is more like sum of endpoints. Currently we mix all endpoint's properties (path segments, parameters, methods) together, but it does not seem to work out nice. I think we have to make endpoints more structured.

@fizruk I've thought about this as well, but I don't want to compromise on ease of use. I think one of the reasons servant "took off" is that aside from `ReqBody`, `Capture` etc, you only have to remember/use the venerable  `:>` and `:<|>`. My current thinking is that we should either figure out a way to keep those but make things nicer, or to base our API descriptions on some precise structure that we can reason about and program in a principled way, and it should not be more complex that the current situation.

**Comment by fizruk** [comment]
@alpmestan 

> I've thought about this as well, but I don't want to compromise on ease of use. I think one of the reasons servant "took off" is that aside from ReqBody, Capture etc, you only have to remember/use the venerable :> and :<|>.

As @kosmikus has mentioned, we don't have to write APIs using the same combinators we use for internal API representation. We can leave `:>` and `:<|>`, just change their definition to be type-level functions (families) that combine (parts of) APIs. Possibly, we could even preserve the type level DSL for APIs exactly as we currently have it.

I agree that Servant took off largely because of its relative simplicity and modularity. I myself bet on it exactly because of this and am now using it commercially for two years! However I think we should start rethinking Servant's API representation and try fix those weird problems we have because of the initial "simple" representation.

**Comment by alpmestan** [comment]
Oh sorry if that wasn't clear but I'm not against changing things. I just wanted to emphasize the fact that whatever we do, we should not make it harder or more complicated for users to write APIs. If anything, it should be simpler or equally simple as it is now, even if we make things fancier internally.

**Comment by phadej** [comment]
@fizruk could you test current `master` with your real stuff; whether the issue is fixed? (no hurry though)

**Comment by fizruk** [comment]
@phadej how do I build against `master`? Tried adding this to `stack.yaml`:

```yaml
- location:
    git: https://github.com/haskell-servant/servant.git
    commit: master
  extra-dep: true
```

But that does not work:

```
$ stack build
Cloning into '.stack-work/downloaded/HHWQO61NVGEM'...
remote: Counting objects: 12511, done.
remote: Compressing objects: 100% (42/42), done.
remote: Total 12511 (delta 14), reused 0 (delta 0), pack-reused 12467
Receiving objects: 100% (12511/12511), 4.49 MiB | 492.00 KiB/s, done.
Resolving deltas: 100% (6719/6719), done.
No .cabal file found in directory .stack-work/downloaded/HHWQO61NVGEM/
```

**Comment by codedmart** [comment]
@fizruk I believe you need to add `subdirs`:
```yaml
- location:
    git: https://github.com/haskell-servant/servant.git
    commit: master
  subdirs:
    - servant
    - servant-server
    - servant-foreign
    - servant-docs
    - servant-client
  extra-dep: true
```

**Comment by fizruk** [comment]
@codedmart thanks!

@phadej I'm not sure it worked...

```
ghc: panic! (the 'impossible' happened)
  (GHC version 8.0.2 for x86_64-apple-darwin):
	Simplifier ticks exhausted
  When trying UnfoldingDone $d~_aIOn
  To increase the limit, use -fsimpl-tick-factor=N (default 100)
  If you need to do this, let GHC HQ know, and what factor you needed
  To see detailed counts use -ddump-simpl-stats
  Total ticks: 220084

Please report this as a GHC bug:  http://www.haskell.org/ghc/reportabug
```

**Comment by phadej** [comment]
@fizruk It might be because of e.g. `aeson`, have you updated to `aeson-1.2` (see changelog: http://hackage.haskell.org/package/aeson-1.2.0.0/changelog).

If that doesn't help, I don't know other way then increasing  the tick factor (and trying to pinpoint the inline offender)

**Comment by fizruk** [comment]
@phadej that is exactly because of the module(s) where I had a workaround for `Enter`.

> trying to pinpoint the inline offender

How can I do that?

**Comment by fizruk** [comment]
@phadej Increasing tick factor to `-fsimpl-tick-factor=1000` does work!
Is it a good solution though?

**Comment by phadej** [comment]
@fizruk it's not good (something big is generated); do you have `{-# INLINE #-}` in your code. I have seen tick exhausting when there was a pragma on something unreasonably big to be inlined?

**Comment by fizruk** [comment]
@phadej nope, I don't have `{-# INLINE #-}`.
I guess it has something to do with your changes to `Enter` and me having a biggish API.

**Comment by phadej** [comment]
@fizruk just guessing, is your monad (you `enter`) type alias to `transformers` stack? They are heavily inlining? Or is it custom monad (`GeneralizedNewtypeDeriving` over `transformers` stack?). 

I'd say that increasing ticks is enough (hopefully it's enough to do only in few files: after all it's safety features). But it would be great to be able to have some small-ish example to report as ghc issue.

**Comment by fizruk** [comment]
@phadej my `App` monad is a newtype over `ReaderT Config Handler`.

> I'd say that increasing ticks is enough (hopefully it's enough to do only in few files: after all it's safety features).

Yeah, thanks for reminding that I can enable options per module!

> But it would be great to be able to have some small-ish example to report as ghc issue.

I'll try to come up with something... maybe this weekend :)
Also, I tried it with 8.0.2, it can be that it's not an issue with 8.2.

**Comment by phadej** [comment]
I'll close this and make new issue for the simplifier ticks exhaustion

**Comment by phadej** [comment]
A comment, if in @fizruk comment we try to do

```haskell
server2 :: App Int :<|> (String -> App Bool)
server2 = enter (NT id) server
```

it will still fail with:

```
    • Overlapping instances for Enter
                                  (App Int :<|> (String -> App Bool))
                                  n0
                                  n0
                                  (App Int :<|> (String -> App Bool))
        arising from a use of ‘enter’
      Matching instances:
        instance (Entered m n (m a) ~ n a, Entered n m (n a) ~ m a) =>
                 Enter (m a) m n (n a)
          -- Defined in ‘Servant.Utils.Enter’
        instance (Enter typ1 m1 n1 ret1, Enter typ2 m2 n2 ret2, m1 ~ m2,
                  n1 ~ n2, Entered m1 n1 (typ1 :<|> typ2) ~ (ret1 :<|> ret2),
                  Entered n1 m1 (ret1 :<|> ret2) ~ (typ1 :<|> typ2)) =>
                 Enter (typ1 :<|> typ2) m1 n1 (ret1 :<|> ret2)
          -- Defined in ‘Servant.Utils.Enter’
      (The choice depends on the instantiation of ‘n0’
       To pick the first instance above, use IncoherentInstances
       when compiling the other instance declarations)
    • In the expression: enter (NT id) server
      In an equation for ‘server2’: server2 = enter (NT id) server
```

I guess we just cannot work-around such situation. But now the error is a bit better.
